### PR TITLE
Add South Africa Crop Type Competition Dataset

### PR DESCRIPTION
**Pull request by GeorgeHuber**
This PR adds the South Africa Crop Type Competition Dataset to TorchGeo. 

[Dataset](https://beta.source.coop/repositories/radiantearth/south-africa-crops-competition/description/)

Similar to https://github.com/microsoft/torchgeo/pull/1459, this dataset does not implement automatic download functions due to issues migrating data from radiantmlhub source cooperative until issue https://github.com/microsoft/torchgeo/issues/1830 is resolved. 

Example field and label in early July:
Screenshot 2024-02-19 at 4 47 04 PM

- [x] Implement the dataset extending either [GeoDataset](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.GeoDataset) or [NonGeoDataset](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.NonGeoDataset)
- [x] Add the dataset definition to torchgeo/datasets/__init__.py
- [x] Add a data.py script to tests/data/<new dataset>/ that generates test data with the same directory structure/file naming conventions as the new dataset
- [x] Add appropriate tests with 100% test coverage to tests/datasets/
- [x] Add the dataset to docs/api/datasets.rst
- [x] Add the dataset metadata to either docs/api/geo_datasets.csv or docs/api/non_geo_datasets.csv

### CODE REVIEW

**Comment by yichiac** [comment]
Please refer to #1459 for removing the download and checksum functions until #1830 is resolved.

**Review by yichiac — CHANGES_REQUESTED** [review]
Please refer to https://github.com/microsoft/torchgeo/pull/1459 for removing the download and checksum functions until https://github.com/microsoft/torchgeo/issues/1830 is resolved.

File: tests/data/south_africa_crop_type/data.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+from rasterio.transform import Affine
+
+
+def generate_test_data() -> str:
+    """Create test data archive for AgriFieldNet dataset.
+    Args:
+        paths: path to store test data
+        n_samples: number of samples.
+    Returns:
+        md5 hash of created archive
```
Comment (adamjstewart) [inline tests/data/south_africa_crop_type/data.py:20]:
```suggestion
    """Create test data archive for SouthAfricaCropType dataset.

    Args:
        paths: path to store test data
        n_samples: number of samples.

    Returns:
        md5 hash of created archive
```

File: tests/data/south_africa_crop_type/data.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            src.write(arr, 1)
+
+    for field_id in train_field_ids:
+        for date in s1_timestamps:
+            s1_dir = os.path.join(train_imagery_s1_dir, field_id, date)
+            os.makedirs(s1_dir, exist_ok=True)
+            for band in s1_bands:
+                train_arr = np.random.randint(
+                    dtype_max, size=(SIZE, SIZE), dtype=dtype
+                )  # noqa: E501
```
Comment (adamjstewart) [inline tests/data/south_africa_crop_type/data.py:79]:
```suggestion
                )
```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+class SouthAfricaCropType(RasterDataset):
+    """South Africa Crop Type Challenge dataset.
+
+    The `South Africa Crop Type Challenge
+    <https://beta.source.coop/repositories/radiantearth/south-africa-crops-competition/description/>`__
+    dataset includes satellite imagery from Sentinel-1 and Sentinel-2 and labels for
+    crop type that were collected by aerial
+    and vehicle survey from May 2017 to March 2018. Data was collected by the
+    provided by the Western Cape Department of Agriculture and is
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:28]: collected by or provided by? Could probably make the length of each line more uniform, it just needs to be less than 88 chars.

File: torchgeo/datasets/south_africa_crop_type.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    is available via the Radiant Earth Foundation.
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:29]: is is available

File: torchgeo/datasets/south_africa_crop_type.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    5: Weeds
+    6: Small grain grazing
+    7: Wheat
+    8: Canola
+    9: Rooibos
+
+    If you use this dataset in your research, please cite the following dataset:
+    Western Cape Department of Agriculture, Radiant Earth Foundation (2021)
+    "Crop Type Classification Dataset for Western Cape, South Africa",
+    Version 1.0, Radiant MLHub, https://doi.org/10.34911/rdnt.j0co8q
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:54]:
```suggestion

    * Western Cape Department of Agriculture, Radiant Earth Foundation (2021)
      "Crop Type Classification Dataset for Western Cape, South Africa",
      Version 1.0, Radiant MLHub, https://doi.org/10.34911/rdnt.j0co8q
```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+    .. versionadded:: 0.6
+    """
+
+    filename_regex = r"""
+        ^(?P<field_id>[0-9]*)_(?P<year>[0-9]{4})_
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:60]: I like to split this on `_` so that it's easier to read. I prefer a single "date" group and `date_format` to match other RasterDatasets. See https://torchgeo.readthedocs.io/en/stable/tutorials/custom_raster_dataset.html for examples.

File: torchgeo/datasets/south_africa_crop_type.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        (?P<month>[0-9]{2})_(?P<day>[0-9]{2})_
+        (?P<band>(B[0-9A-Z]{2} | VH | VV))_10m\.tif"""
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:62]: I usually don't add a suffix like `.tif` in case someone wants to convert the file to a different format.

File: torchgeo/datasets/south_africa_crop_type.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    all_bands: list[str] = S1_bands + S2_bands
+    cmap = {
+        0: (0, 0, 0, 255),
+        1: (255, 211, 0, 255),
+        2: (255, 37, 37, 255),
+        3: (0, 168, 226, 255),
+        4: (255, 158, 9, 255),
+        5: (37, 111, 0, 255),
+        6: (255, 255, 0, 255),
+        8: (111, 166, 0, 255),
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:89]: Are you missing 7?

File: torchgeo/datasets/south_africa_crop_type.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        9: (0, 175, 73, 255),
+    }
+
+    def __init__(
+        self,
+        root: str = "",
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:95]:
```suggestion
        root: str = "data",
```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        crs: CRS = CRS.from_epsg(32634),
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:96]: We don't normally specify a default CRS unless one is needed.

File: torchgeo/datasets/south_africa_crop_type.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        classes: list[int] = list(cmap.keys()),
+        bands: list[str] = all_bands,
+        transforms: Optional[Callable[[dict[str, Tensor]], dict[str, Tensor]]] = None,
+    ) -> None:
+        """Initialize a new South Africa dataset instance.
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:101]:
```suggestion
        """Initialize a new South Africa Crop Type dataset instance.
```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+        Args:
+            root: root directory where dataset can be found
+            crs: coordinate reference system to be used
+            classes: crop type classes to be included
+            bands: the subset of bands to load
+            transforms: a function/transform that takes input sample and its target as
+                entry and returns a transformed version
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:109]:
```suggestion
                entry and returns a transformed version

```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+            fields_added = []
+            band_filepaths = []
+            for filepath in filepaths:
+                filename = os.path.basename(filepath)
+                match = re.match(filename_regex, filename)
+                if match and match.groupdict()["band"] == band:
+                    field_id = match.groupdict()["field_id"]
+                    month = match.groupdict()["month"]
+                    day = match.groupdict()["day"]
+                    if field_id not in fields_added and month == "07":
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:163]: What's the deal with this? Seems like there are a lot of custom hacks with no comments to explain them.

File: torchgeo/datasets/south_africa_crop_type.py (hunk 9 above)
Comment (adamjstewart) [reply torchgeo/datasets/south_africa_crop_type.py:95]: Shouldn't this be `paths` to match all other GeoDatasets?

File: torchgeo/datasets/south_africa_crop_type.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+            raise IndexError(
+                f"query: {query} not found in index with bounds: {self.bounds}"
+            )
+
+        data_list: list[Tensor] = []
+        filename_regex = re.compile(self.filename_regex, re.VERBOSE)
+
+        sample_dates_s1: dict[str, str] = {}
+        sample_dates_s2: dict[str, str] = {}
+        for band in self.bands:
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:152]: Remind me again why we have to copy all of this instead of just calling `super().__getitem__(...)`?

File: torchgeo/datasets/south_africa_crop_type.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+                        if field_id not in sample_dates:
+                            sample_dates[field_id] = day
+                        if day == sample_dates[field_id]:
+                            band_filepaths.append(filepath)
+                            fields_added.append(field_id)
+            data_list.append(self._merge_files(band_filepaths, query))
+        image = torch.cat(data_list)
+
+        mask_filepaths = []
+        for root, dirs, files in os.walk(os.path.join(self.root, "train", "labels")):
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:173]: Seems like this would result in the mask containing every file in the dataset every time. Is this correct? Seems incredibly inefficient to do this every time `__getitem__` is called.

File: torchgeo/datasets/south_africa_crop_type.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    0: No Data
+    1: Lucerne/Medics
+    2: Planted pastures (perennial)
+    3: Fallow
+    4: Wine grapes
+    5: Weeds
+    6: Small grain grazing
+    7: Wheat
+    8: Canola
+    9: Rooibos
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:49]:
```suggestion
    0. No Data
    1. Lucerne/Medics
    2. Planted pastures (perennial)
    3. Fallow
    4. Wine grapes
    5. Weeds
    6. Small grain grazing
    7. Wheat
    8. Canola
    9. Rooibos
```
Needed for correct list formatting by Sphinx

File: torchgeo/datasets/south_africa_crop_type.py [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+    """
+
+    filename_regex = r"""
+        ^(?P<field_id>[0-9]*)_(?P<year>[0-9]{4})_
+        (?P<month>[0-9]{2})_(?P<day>[0-9]{2})_
+        (?P<band>(B[0-9A-Z]{2} | VH | VV))_10m\.tif"""
+
+    rgb_bands = ["B04", "B03", "B02"]
+    S1_bands = ["VH", "VV"]
+    S2_bands = [
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:66]: Let's use a lowercase s. See PEP-8.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/datasets/south_africa_crop_type.py [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+    <https://beta.source.coop/repositories/radiantearth/south-africa-crops-competition/description/>`__
+    dataset includes satellite imagery from Sentinel-1 and Sentinel-2 and labels for
+    crop type that were collected by aerial and vehicle survey from May 2017 to March
+    2018. Data was provided by the Western Cape Department of Agriculture and is
+    available via the Radiant Earth Foundation. Each chip is matched with a label.
+    Each pixel in the label contains an integer field number and crop type class.
+
+    Dataset format:
+
+    * images are 2-band Sentinel 1 and 12-bands Sentinel-2 data with a cloud mask
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:34]:
```suggestion
    * images are 2-band Sentinel 1 and 12-band Sentinel-2 data with a cloud mask
```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+            self.ordinal_cmap[v] = torch.tensor(self.cmap[k])
+
+    def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            data adn labels at that index
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:140]:
```suggestion
            data and labels at that index
```

**Review by adamjstewart — COMMENTED** [review]
I don't see any remaining major issues. Just wondering if `__getitem__` can be further simplified. Also need comments in `__getitem__` to explain all the non-standard stuff.

**Comment by GeorgeHuber** [comment]
> I don't see any remaining major issues. Just wondering if `__getitem__` can be further simplified. Also need comments in `__getitem__` to explain all the non-standard stuff.

I added some comments to explain the extra stuff, let me know if you think it needs more detail. The multiple images to one label thing is really where most of the complication comes from along with the need to make sure the labels and fields are in the same order since they won't be sorted alphabetically. I thought this would be the most efficient solution as we loop over the hit filepaths once just as we would standardly but I'm open to other ideas.

File: torchgeo/datasets/south_africa_crop_type.py [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+class SouthAfricaCropType(RasterDataset):
+    """South Africa Crop Type Challenge dataset.
+
+    The `South Africa Crop Type Challenge
+    <https://beta.source.coop/repositories/radiantearth/south-africa-crops-competition/description/>`__
+    dataset includes satellite imagery from Sentinel-1 and Sentinel-2 and labels for
+    crop type that were collected by aerial and vehicle survey from May 2017 to March
+    2018. Data was provided by the Western Cape Department of Agriculture and is
+    available via the Radiant Earth Foundation. Each chip is matched with a label.
+    Each pixel in the label contains an integer field number and crop type class.
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:30]: Maybe worth adding a note saying that the dataset contains time-series data, but that this is not yet exposed, and explain which time will be returned

File: torchgeo/datasets/south_africa_crop_type.py [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+
+    Dataset format:
+
+    * images are 2-band Sentinel 1 and 12-band Sentinel-2 data with a cloud mask
+    * masks are tiff image with unique values representing the class and field id.
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:35]:
```suggestion
    * masks are tiff images with unique values representing the class and field id.
```

File: torchgeo/datasets/south_africa_crop_type.py [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+        # Loop through matched filepaths and find all unique field ids
+        field_ids: list[str] = []
+        # Store date in July for s1 and s2 we want to use for each sample
+        imagery_dates: dict[str, dict[str, str]] = {}
+
+        for filepath in filepaths:
+            filename = os.path.basename(filepath)
+            match = re.match(filename_regex, filename)
+            if match:
+                field_id = match.groupdict()["field_id"]
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:165]:
```suggestion
                field_id = match.group("field_id")
```
this is simpler

**Review by adamjstewart — COMMENTED** [review]
Probably worth

File: torchgeo/datasets/south_africa_crop_type.py [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+    def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            data and labels at that index
+        """
+        assert isinstance(self.paths, str)
```
Comment (adamjstewart) [inline torchgeo/datasets/south_africa_crop_type.py:146]: It's probably possible to relax this constraint by using `filepaths` to compute the correct directory for the mask. But let's do that another day, I want to get this merged.

**Review by adamjstewart — APPROVED** [review]
